src/frontend/src/optimizer/plan_node/logical_agg.rs

```diff
@@ -242,6 +290,91 @@ impl LogicalAgg {
         }
     }
 
+    fn separate_normal_and_special_agg(&self) -> SeparatedAggInfo {
+        let estimated_len = self.agg_calls().len() - 1;
+        let mut approx_percentile_agg_calls = Vec::with_capacity(estimated_len);
+        let mut non_approx_percentile_agg_calls = Vec::with_capacity(estimated_len);
+        let mut approx_percentile_col_mapping = Vec::with_capacity(estimated_len);
+        let mut non_approx_percentile_col_mapping = Vec::with_capacity(estimated_len);
+        for (output_idx, agg_call) in self.agg_calls().iter().enumerate() {
+            if agg_call.agg_kind == AggKind::Builtin(PbAggKind::ApproxPercentile) {
+                approx_percentile_agg_calls.push(agg_call.clone());
+                approx_percentile_col_mapping.push(Some(output_idx));
+            } else {
+                non_approx_percentile_agg_calls.push(agg_call.clone());
+                non_approx_percentile_col_mapping.push(Some(output_idx));
+            }
+        }
+        SeparatedAggInfo {
+            approx_percentile_agg_calls,
+            non_approx_percentile_agg_calls,
+            approx_percentile_col_mapping: ColIndexMapping::new(
+                approx_percentile_col_mapping,
+                self.agg_calls().len(),
+            ),
+            non_approx_percentile_col_mapping: ColIndexMapping::new(
+                non_approx_percentile_col_mapping,
+                self.agg_calls().len(),
+            ),
+        }
+    }
+
+    fn build_approx_percentile_agg(
+        &self,
+        input: PlanRef,
+        approx_percentile_agg_call: &PlanAggCall,
+    ) -> PlanRef {
+        let local_approx_percentile =
+            StreamLocalApproxPercentile::new(input, approx_percentile_agg_call);
+        let global_approx_percentile = StreamGlobalApproxPercentile::new(
+            local_approx_percentile.into(),
+            approx_percentile_agg_call,
+        );
+        global_approx_percentile.into()
+    }
```

Comment:
Is it possible to unify `ApproxPercentile` executor with `ApproxCountDistinct` to have something like `BucketBasedAgg`? Just curious, not a suggestion.

Reply:
Hmm I'm not sure about the internals of approx count distinct. Does it involve 2 phase agg?

Reply:
No I mean is it possible to abstract a common executor like HashAgg that can do various kinds of aggregation that can fit into the framework. This seems not related to 2-phase agg, cuz HashAgg can be used in either local phase or global phase.